src/repr/src/row.rs

```diff
@@ -1256,17 +1247,18 @@ impl RowArena {
     /// would be called `push_owned_datum`.
     pub fn push_unary_row<'a>(&'a self, row: Row) -> Datum<'a> {
         let mut inner = self.inner.borrow_mut();
-        inner.owned_rows.push(row);
-        let datum = inner.owned_rows[inner.owned_rows.len() - 1].unpack_first();
+        inner.push(row.data.into_vec());
         unsafe {
             // This is safe because:
-            //   * We only ever append to self.owned_rows, so the row will live
+            //   * We only ever append to self.inner, so the row data will live
             //     as long as the arena.
-            //   * We return a reference to the heap allocation inside the row,
-            //     so it's okay if self.owned_rows reallocates and moves the
-            //     row.
-            //   * We don't allow access to the row itself, so row.data will
+            //   * We force the row data into its own heap allocation--
+            //     importantly, we do NOT store the SmallVec, which might be
+            //     storing data inline on the stack--so it's okay if self.inner=
+            //     reallocates and moves the row.
```

Comment:
Just to check my understanding, is the problem stack storage, or that when the vector resizes it may invalidate the (heap allocated) `SmallVec` itself, when it moves it to a new allocation?

Reply:
Oops, I got this right in the commit message but not here. Sorry, it's late. The issue was that if the `SmallVec` was using inline storage, that inline storage would move around as the outer `Vec` reallocated. And we were just happily handing out a pointer to that moveable inline storage. That data wasn't actually on the stack; it was just part of the outer heap allocation instead of an independent heap allocation.

Reply:
That makes sense, and I just wanted to make sure I understood. Thanks!